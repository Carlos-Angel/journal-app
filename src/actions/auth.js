import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
} from 'firebase/auth';
import { googleAuthProvider } from 'app';
import { authTypes } from 'types';

export const startLoginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'pedro'));
    }, 3500);
  };
};

export const startRegister = ({ email, password, name }) => {
  return (dispatch) => {
    const auth = getAuth();
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => console.error(error.message));
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch((error) => console.error(error.message));
  };
};

export const login = (uid, displayName) => {
  return {
    type: authTypes.login,
    payload: {
      uid,
      displayName,
    },
  };
};
