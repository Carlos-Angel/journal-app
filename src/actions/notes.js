import Swal from 'sweetalert2';
import { db } from 'app';
import { notesTypes } from 'types';
import {
  addDoc,
  collection,
  updateDoc,
  doc,
  deleteDoc,
} from 'firebase/firestore';
import { loadNotes } from 'helpers/loadNotes';
import { fileUpload } from 'helpers/fileUpload';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const note = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const document = await addDoc(collection(db, `${uid}/journal/notes`), note);
    dispatch(activeNote(document.id, note));
    dispatch(addNote(document.id, note));
  };
};

export const activeNote = (id, note) => ({
  type: notesTypes.active,
  payload: {
    id,
    ...note,
  },
});

export const addNote = (id, note) => ({
  type: notesTypes.add,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: notesTypes.load,
  payload: notes,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    try {
      const referenceDocument = doc(db, uid, 'journal', 'notes', note.id);
      await updateDoc(referenceDocument, noteToFirestore);
      dispatch(refreshNote(note.id, noteToFirestore));
      Swal.fire('Saved', note.title, 'success');
    } catch (error) {
      Swal.fire('Error', 'error al guardar la nota', 'error');
    }
  };
};

export const refreshNote = (id, note) => ({
  type: notesTypes.update,
  payload: {
    id,
    ...note,
  },
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const referenceDocument = doc(db, `${uid}/journal/notes/${id}`);
    await deleteDoc(referenceDocument);
    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => ({
  type: notesTypes.delete,
  payload: id,
});

export const deleteNotes = () => ({
  type: notesTypes.logoutCleaning,
});
