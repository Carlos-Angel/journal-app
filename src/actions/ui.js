import { authTypes } from 'types';

export const setError = (message) => {
  return {
    type: authTypes.uiSetError,
    payload: message,
  };
};

export const removeError = () => {
  return {
    type: authTypes.uiRemoveError,
  };
};
