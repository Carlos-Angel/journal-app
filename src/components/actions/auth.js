import { authTypes } from '../types';

export const startLoginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    setTimeout(() => {
      dispatch(login(123, 'pedro'));
    }, 3500);
  };
};

export const login = (uuid, displayName) => {
  return {
    type: authTypes.login,
    payload: {
      uuid,
      displayName,
    },
  };
};
