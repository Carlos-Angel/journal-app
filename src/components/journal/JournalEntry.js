import React from 'react';

export const JournalEntry = () => {
  return (
    <div className='journal__entry pointer'>
      <div
        className='journal__entry-picture'
        style={{
          backgroundSize: 'cover',
          backgroundImage:
            'url(https://enfocando.es/wp-content/2016/08/Panorama_7490.jpg)',
        }}
      />
      <div className='journal__entry-body'>
        <p className='journal__entry-title'>titulo del journal</p>
        <p className='journal__entry-content'>
          Lorem ipsum dolor, sit amet consectetur adipisicing elit.
        </p>
      </div>
      <div className='journal__entry-date-box'>
        <span>Monday</span>
        <h4>28</h4>
      </div>
    </div>
  );
};
