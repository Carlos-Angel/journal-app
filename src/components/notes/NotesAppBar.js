import { startSaveNote, startUploading } from 'actions/notes';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

export const NotesAppBar = () => {
  const { active: note } = useSelector((state) => state.notes);
  const dispatch = useDispatch();
  const handleSave = () => {
    dispatch(startSaveNote(note));
  };

  const handlePictureClick = () => {
    console.log('handlePictureClick');
    document.querySelector('#fileSelector').click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploading(file));
    }
  };

  return (
    <div className='notes__appbar'>
      <span>25 de agosto del 2021</span>
      <input
        type='file'
        name='file'
        style={{ display: 'none' }}
        onChange={handleFileChange}
        id='fileSelector'
      />
      <div>
        <button className='btn' onClick={handlePictureClick}>
          picture
        </button>
        <button className='btn' onClick={handleSave}>
          save
        </button>
      </div>
    </div>
  );
};
