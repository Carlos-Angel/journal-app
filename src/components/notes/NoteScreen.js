import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NotesAppBar } from './NotesAppBar';
import { useForm } from 'hooks/useForm';
import { activeNote } from 'actions/notes';

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);
  const [{ body, title }, handleInputChange, reset] = useForm(note);
  const activeId = useRef(note.id);
  const dispatch = useDispatch();

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(note.id, { ...note, title, body }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [body, title]);

  return (
    <div className='notes__main-content'>
      <NotesAppBar />

      <div className='notes__content'>
        <input
          type='text'
          placeholder='Some awesome title'
          name='title'
          className='notes__title-input'
          autoComplete='off'
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          name='body'
          className='notes__textarea'
          placeholder='what happened today'
          value={body}
          onChange={handleInputChange}
        ></textarea>

        <div className='notes__image'>
          {note?.url && <img src={note.url} alt='title-note' />}
        </div>
      </div>
    </div>
  );
};
