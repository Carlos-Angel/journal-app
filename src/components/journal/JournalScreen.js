import React from 'react';
import { NoteScreen } from 'components/notes/NoteScreen';
import { NothingSelected } from './NothingSelected';
import { Sidebar } from './Sidebar';

export const JournalScreen = () => {
  return (
    <div className='journal__main-content'>
      <Sidebar />
      <main>{true ? <NoteScreen /> : <NothingSelected />}</main>
    </div>
  );
};
