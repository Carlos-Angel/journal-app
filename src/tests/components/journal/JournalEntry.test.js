import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { activeNote } from 'actions/notes';
import { JournalEntry } from 'components/journal/JournalEntry';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

describe('pruebas en <JournalEntry />', () => {
  beforeEach(() => {
    store.clearActions();
    jest.clearAllMocks();
  });

  const note = {
    id: '1',
    date: 0,
    title: 'title note',
    body: 'body note',
    url: 'https://imagen.com/test.jpg',
  };

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <JournalEntry {...note} />
      </MemoryRouter>
    </Provider>,
  );

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de activar la nota', () => {
    wrapper.find('.journal__entry').prop('onClick')();

    expect(store.dispatch).toHaveBeenCalledWith(
      activeNote(note.id, { ...note }),
    );
  });
});
