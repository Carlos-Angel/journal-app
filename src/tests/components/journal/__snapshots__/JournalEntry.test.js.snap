// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <JournalEntry /> debe de mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url('https://imagen.com/test.jpg')",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      title note
    </p>
    <p
      className="journal__entry-content"
    >
      body note
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
